### PR TITLE
calculateSetParametersAsync: Precision calculator based on maxPercentError

### PR DESCRIPTION
**Pull request by felix2feng**
- Allows the user to input a maximum percent error.
- The function will attempt to calculate a natural unit / component unit combination that minimizes the error between the implied set price and the user-supplied target set price.

### CODE REVIEW

File: src/api/FactoryAPI.ts [hunk 1]
```diff
@@ -116,24 +117,45 @@ export class FactoryAPI {
...
+          .div(ether(1))
+          .ceil();
+      });
+
+      const impliedSetPrice = this.calculateImpliedSetPrice(formattedComponentUnits, naturalUnit, prices);
+      percentError = calculatePercentDifference(impliedSetPrice, targetPrice);
+
+      console.log(percentError.toString(), maxPercentError.toString(), priorPercentError.toString());
+
+      // Only continue to experiemnt with improvements if the following conditions are met:
```
Comment (justinkchen) [inline src/api/FactoryAPI.ts]: `experiment`

File: src/api/FactoryAPI.ts [hunk 2]
```diff
@@ -116,24 +117,45 @@ export class FactoryAPI {
...
+        return amountRequired
+          .mul(naturalUnit)
+          .div(ether(1))
+          .ceil();
+      });
+
+      const impliedSetPrice = this.calculateImpliedSetPrice(formattedComponentUnits, naturalUnit, prices);
+      percentError = calculatePercentDifference(impliedSetPrice, targetPrice);
+
+      console.log(percentError.toString(), maxPercentError.toString(), priorPercentError.toString());
```
Comment (justinkchen) [inline src/api/FactoryAPI.ts]: Remove this

File: src/SetProtocol.ts [hunk 3]
```diff
@@ -110,27 +110,27 @@ class SetProtocol {
...
+   * @return                  Object conforming to NewSetParameters containing a list of component units in index order
    *                           and a valid natural unit. These can be passed directly into `createSetAsync`
    */
   public async calculateSetParametersAsync(
     components: Address[],
     prices: BigNumber[],
     proportions: BigNumber[],
     targetPrice: BigNumber,
-    precision: BigNumber = new BigNumber(0),
+    maxPercentError: BigNumber = new BigNumber(1),
```
Comment (asoong) [inline src/SetProtocol.ts]: `percentAllowableError`

File: src/SetProtocol.ts [hunk 4]
```diff
@@ -110,27 +110,27 @@ class SetProtocol {
    * Calculates inputs for `createSetAsync` for a given list of ERC20 token addreses, proportions of each,
```
Comment (asoong) [inline src/SetProtocol.ts]: Add a blurb here about what the precision is for? We do this in `FactoryAPI#createSetAsync` for example

File: src/api/FactoryAPI.ts [hunk 5]
```diff
@@ -116,24 +117,43 @@ export class FactoryAPI {
...
-    return {
-      units: formattedComponentUnits,
-      naturalUnit,
-    };
+    let naturalUnit = BigNumber.max(minimumNaturalUnit, derivedNaturalUnit);
+    let formattedComponentUnits: BigNumber[];
+    let priorPercentError: BigNumber = E18; // Start with a large percentage figure
+    let percentError: BigNumber;
+
+    // If the percentage error from the naturalUnit and units combination is greater
```
Comment (asoong) [inline src/api/FactoryAPI.ts]: I like this description, can reuse something like this!

File: src/constants/index.ts [hunk 6]
```diff
@@ -34,6 +34,11 @@ export const STANDARD_TRANSFER_VALUE: BigNumber = new BigNumber(1000000000000000
 export const UNLIMITED_ALLOWANCE_IN_BASE_UNITS: BigNumber =
   SetProtocolUtils.CONSTANTS.UNLIMITED_ALLOWANCE_IN_BASE_UNITS; // 2 ** 256 - 1
 export const ZERO: BigNumber = SetProtocolUtils.CONSTANTS.ZERO;
+export const ONE: BigNumber = new BigNumber(1);
+export const TWO: BigNumber = new BigNumber(2);
+export const TEN: BigNumber = new BigNumber(10);
+export const EIGHTEEN: BigNumber = new BigNumber(18);
+export const E18: BigNumber = new BigNumber(10).pow(18);
```
Comment (asoong) [inline src/constants/index.ts]: Can we name this something else? 1 ether in wei?

File: src/constants/index.ts [hunk 7]
```diff
@@ -34,6 +34,11 @@ export const STANDARD_TRANSFER_VALUE: BigNumber = new BigNumber(1000000000000000
 export const UNLIMITED_ALLOWANCE_IN_BASE_UNITS: BigNumber =
   SetProtocolUtils.CONSTANTS.UNLIMITED_ALLOWANCE_IN_BASE_UNITS; // 2 ** 256 - 1
 export const ZERO: BigNumber = SetProtocolUtils.CONSTANTS.ZERO;
+export const ONE: BigNumber = new BigNumber(1);
+export const TWO: BigNumber = new BigNumber(2);
+export const TEN: BigNumber = new BigNumber(10);
```
Comment (asoong) [inline src/constants/index.ts]: I get what you're doing, it's awkward to define a big number in line and then use it immediately, but I wasn't sure what/why this constant existed. I'd actually prefer what we did in contracts and have an `ether(n)` function. Maybe something like `transactionUsableNumber(n)`, or maybe `uint256(n)`?

**Pull request merged**